pkg/epoll/epoll.go

```diff
@@ -0,0 +1,211 @@
+/*
+Copyright 2020 The OpenEBS Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package epoll
+
+import (
+	"errors"
+	"os"
+	"syscall"
+	"time"
+)
+
+const (
+	EPOLLIN  EventType = syscall.EPOLLIN
+	EPOLLPRI EventType = syscall.EPOLLPRI
+	EPOLLERR EventType = syscall.EPOLLERR
+	EPOLLOUT EventType = syscall.EPOLLOUT
+	EPOLLHUP EventType = syscall.EPOLLHUP
+
+	// timeout is the duration after which the EpollWait system
+	// call returns even if there are no events.
+	timeout int = int(time.Millisecond)
+)
+
+var (
+	ErrFileAlreadyWatched error = errors.New("file being watched already")
+	ErrInvalidEventType   error = errors.New("invalid event type")
+	ErrWatcherNotFound    error = errors.New("watcher not found")
+)
+
+type Epoll struct {
+	watchers      []Watcher
+	fileToWatcher map[string]*Watcher
+	fdToWatcher   map[int32]*Watcher
```

Comment:
Do we need both FD and file to watchers?

Reply:
I added them so that I don't have to iterate through the watcher slice again and again.

Reply:
Both are needed.
1. When adding/deleting a watcher we search for watcher based on the file name.
2. When syscall.EpollWait returns any events, the events are mapped to fd. so we need to search for the watcher based on the fd.